### PR TITLE
Add flake8 to Travis

### LINKED ISSUE #24: Add flake8 checks
We should run `flake8` against the code in Travis to ensure we're following good coding guidelines. Linting should also happen before the code is tested to save time.

### PR DESCRIPTION
**Pull request by major**
This patch set adds flake8 testing in the Travis job and also runs
the linting checks prior to the python tests.

Fixes for flake8 issues are included.

Depends on #22 to merge first.
Fixes #24.

Signed-off-by: Major Hayden <major@redhat.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,6 +31,34 @@ For running some specific tests you can do this as following:
...
+    -h, --help            show this help message and exit
+    -d WORKDIR, --workdir WORKDIR
+                            Path to work dir
+    -w, --wipe            Clean build (make mrproper before building), remove
+                            workdir when finished
+    --junit JUNIT         Path to dir to store junit results in
+    -v, --verbose         Increase verbosity level
+    --rc RC               Path to rc file
+    --state               Save/read state from 'state' section of rc file
+
```
Comment (spbnick) [inline README.md]: Could we perhaps not include the `skt -h` output here, and e.g. simply say that `skt -h` should produce command-line help, or such? This way README.md would be shorter and we won't have to update it every time we change the `skt -h` output.

File: setup.cfg [hunk 2]
```diff
@@ -0,0 +1,19 @@
+[metadata]
+name = skt
+description = "Sonic Kernel Testing"
+long_description = file: README.md
+version = 0.0.1
```
Comment (spbnick) [inline setup.cfg]: Would you be open to just having `version = 1` here? I'm aware of semantic versioning, but, I find that so many numbers in version are not needed until at least half a dozen releases. Meanwhile we'll be having major changes in the code and version number changes won't reflect that. We can always add dots and number when we need them.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+"""Install skt using setuptools."""
+from setuptools import setup
+
+
+setup()
```
Comment (spbnick) [inline setup.py]: I don't know how Python packaging works, sorry, but does this file have to have `.py` in the name?

File: skt/executable.py [hunk 4]
```diff
@@ -19,16 +19,18 @@
...
 import sys
 import time
+
+import junit_xml
+
+import skt
 import skt.publisher
 import skt.reporter
 import skt.runner
-import skt
```
Comment (spbnick) [inline skt/executable.py]: I'm curious, what's the logic behind the import reordering you're doing here and other files? Could you perhaps put those into a separate commit and elaborate there?

File: skt/executable.py [hunk 5]
```diff
@@ -123,7 +125,7 @@ def cmd_merge(cfg):
         idx = 0
         for mb in cfg.get('merge_ref'):
             save_state(cfg, {'meregerepo_%02d' % idx: mb[0],
-                             'mergehead_%02d' % idx: head})
+                             'mergehead_%02d' % idx: bhead})
```
Comment (spbnick) [inline skt/executable.py]: Hmm, could you please make this a separate commit and elaborate a little bit on what's going on? Thanks!

File: skt/__init__.py [hunk 6]
```diff
@@ -37,6 +37,11 @@ def stringify(v):
         return str(v)
 
 
+class RpcProtocolMismatch(Exception):
+    """Exception handler for Patchwork API mismatch errors."""
+    pass
+
+
```
Comment (spbnick) [inline skt/__init__.py]: Hmm, did you intend to use this somewhere and forget?

File: README.md [hunk 7]
```diff
@@ -82,7 +110,7 @@ default.
 
 E.g. to build with the current system's config file run:
 
-    $ skt.py --workdir skt-workdir -vv \
+    $ skt --workdir skt-workdir -vv \
              build -c /boot/config-`uname -r`
```
Comment (spbnick) [inline README.md]: Ah, those feel nice :relieved:.

File: .travis.yml [hunk 8]
```diff
@@ -4,6 +4,7 @@ python:
 install:
     - pip install requests junit_xml beaker-client mock
     - pip install pylint pycodestyle
+    - pip install .
```
Comment (spbnick) [inline .travis.yml]: Shouldn't this be a part of tests instead, i.e. be in the `script` list? And should we perhaps add an uninstall test as well?

File: .travis.yml [hunk 9]
```diff
@@ -3,9 +3,11 @@ python:
     - "2.7"
 install:
     - pip install requests junit_xml beaker-client mock
-    - pip install pylint pycodestyle
+    - pip install pylint pycodestyle flake8
     - pip install .
+before_script:
+    - flake8 --show-source .
+    - pycodestyle .
```
Comment (spbnick) [inline .travis.yml]: Travis docs [say](https://docs.travis-ci.com/user/customizing-the-build/#Breaking-the-Build) that any failure in `before_script` aborts the whole "build", i.e. none of the `script` commands would run. Was that your intention?

**Comment by spbnick** [comment]
Ah, one thing I missed: what would be the procedure to run skt from the source tree after this change?

File: .travis.yml (hunk 8 above)
Comment (major) [reply .travis.yml]: Sounds reasonable. I'll make those changes.

File: README.md (hunk 1 above)
Comment (major) [reply README.md]: That works for me.

File: setup.cfg (hunk 2 above)
Comment (major) [reply setup.cfg]: Sure. ;)

File: setup.py (hunk 3 above)
Comment (major) [reply setup.py]: Yes, the `.py` is required because pip/setuptools look for `setup.py` and `setup.cfg` (among other files) when installing a new python module.

**Comment by major** [comment]
Waiting for #22 to merge first...

File: .travis.yml [hunk 10]
```diff
@@ -3,8 +3,12 @@ python:
     - "2.7"
 install:
     - pip install requests junit_xml beaker-client mock
-    - pip install pylint pycodestyle
+    - pip install pylint pycodestyle flake8
+before_script:
+    - flake8 --show-source .
+    - pycodestyle .
```
Comment (agustinhenze) [inline .travis.yml]: Quoting the flake8's doc:
```
Flake8 is a wrapper around PyFlakes, pycodestyle, and Ned's McCabe script.
Flake8 runs all tools by launching the single 'flake8' script, but ignores pycodestyle and PyFlakes extended options and just uses defaults.
```
So I think that you should remove `pycodestyle .` item.

Comment (major) [reply .travis.yml]: Ah, good catch, @agustinhenze. I'll fix that now.

File: skt/__init__.py (hunk 6 above)
Comment (major) [reply skt/__init__.py]: This exception was referenced in the code already, but it didn't exist. That caused a linting error.

File: skt/executable.py (hunk 4 above)
Comment (major) [reply skt/executable.py]: That is being done to please `flake8`. There are ordering and spacing requirements for modules that come from core python, extra modules installed later, and the skt project itself.

Comment (spbnick) [reply skt/executable.py]: Wow, that's heavy. And I thought *I* was a perfectionist :smiley:.

File: skt/__init__.py (hunk 6 above)
Comment (spbnick) [reply skt/__init__.py]: Riight. Then we need the same change in sktm, because it has the same code there (we'll also probably need a library shared by skt and sktm).

Could you fix the description, though? This is not an "Exception handler", just an "Exception". Also, please don't put a full stop after a single-sentence docstring.

File: skt/executable.py [hunk 11]
```diff
@@ -124,7 +124,7 @@ def cmd_merge(cfg):
     try:
         idx = 0
         for mb in cfg.get('merge_ref'):
-            save_state(cfg, {'meregerepo_%02d' % idx: mb[0],
+            save_state(cfg, {'mergerepo_%02d' % idx: mb[0],
```
Comment (spbnick) [inline skt/executable.py]: Hey, good catch!

File: .travis.yml [hunk 12]
```diff
@@ -3,10 +3,11 @@ python:
...
 install:
     - pip install requests junit_xml beaker-client mock
-    - pip install pylint pycodestyle
+    - pip install pylint pycodestyle flake8
+before_script:
+    - flake8 --show-source .
 script:
     - python -m unittest discover tests
     - pylint tests
-    - pycodestyle .
```
Comment (spbnick) [inline .travis.yml]: Does flake8 include pycodestyle tests, all of them?

File: .travis.yml [hunk 13]
```diff
@@ -3,10 +3,11 @@ python:
     - "2.7"
 install:
     - pip install requests junit_xml beaker-client mock
-    - pip install pylint pycodestyle
+    - pip install pylint pycodestyle flake8
+before_script:
+    - flake8 --show-source .
```
Comment (spbnick) [inline .travis.yml]: Hmm, this means that nothing from `script` would run, if this fails. Was that your intention?

File: .travis.yml (hunk 12 above)
Comment (major) [reply .travis.yml]: Yes. @agustinhenze pointed that out to me in another PR. Hooray for less code! :smile:

File: .travis.yml (hunk 13 above)
Comment (major) [reply .travis.yml]: Is `script` a directory somewhere? I'm not sure where that might be.

Running `flake8` like this is recommended in their docs and it's a common practice when using `tox` for python testing, too.

**Comment by major** [comment]
@spbnick Just pushed a fix for the exception issue you noted.

File: skt/__init__.py [hunk 14]
```diff
@@ -37,6 +37,11 @@ def stringify(v):
         return str(v)
 
 
+class RpcProtocolMismatch(Exception):
+    """Exception for Patchwork API mismatch errors"""
+    pass
```
Comment (agustinhenze) [inline skt/__init__.py]: `pass` is not needed here

Comment (major) [reply skt/__init__.py]: Fixed in latest push. ;)

File: skt/__init__.py (hunk 6 above)
Comment (spbnick) [reply skt/__init__.py]: I think this should be in the first patch of the PR.

File: .travis.yml (hunk 12 above)
Comment (spbnick) [reply .travis.yml]: Right, then we'll need to remove `pycodestyle` from [CODING.md](https://github.com/RH-FMK/meta/blob/master/CODING.md), once we switch sktm to flake8.

File: .travis.yml (hunk 13 above)
Comment (spbnick) [reply .travis.yml]: No, I meant `script` array next to `before_script` in `.travis.yml`. Travis CI doc [says](https://docs.travis-ci.com/user/customizing-the-build/#Breaking-the-Build) if any of the `before_script` commands fails, nothing under `script` runs, or so I understood. I'm not sure why flake8 recommends that, but perhaps because they think it's very important. I'd say being able to also get test results in a single CI run is a good thing, so I would move that to `script`.

**Comment by major** [comment]
@spbnick Sorry for putting that `pass` removal in the last commit. I guess I was in a hurry and squashed it into that last one. ;)

**Comment by spbnick** [comment]
Thanks a lot, Major, merged!
